flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java

```diff
@@ -306,51 +306,88 @@ public void testProducerFailedException() throws Exception {
 
-			inputChannel.onSenderBacklog(11);
+			// Increase the backlog to exceed the number of available floating buffers
+			inputChannel.onSenderBacklog(10);
 
-			// Need extra three floating buffers, but only two buffers available in buffer pool, register as listener as a result
-			verify(bufferPool, times(11)).requestBuffer();
+			// The channel does not get enough floating buffer and register as buffer listener
+			verify(bufferPool, times(13)).requestBuffer();
 			verify(bufferPool, times(1)).addBufferListener(inputChannel);
-			assertEquals("There should be 12 buffers available in the channel",
-				12, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 11 buffers available in the channel", 11, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 12 buffers required in the channel", 12, inputChannel.getNumberOfRequiredBuffers());
+			assertEquals("There should be 0 buffer available in local pool", 0, bufferPool.getNumberOfAvailableMemorySegments());
 
-			inputChannel.onSenderBacklog(12);
+			// Continue increasing the backlog
+			inputChannel.onSenderBacklog(11);
 
-			// Already in the status of waiting for buffers and will not request any more
-			verify(bufferPool, times(11)).requestBuffer();
+			// The channel is already in the status of waiting for buffers and will not request any more
+			verify(bufferPool, times(13)).requestBuffer();
 			verify(bufferPool, times(1)).addBufferListener(inputChannel);
+			assertEquals("There should be 11 buffers available in the channel", 11, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 13 buffers required in the channel", 13, inputChannel.getNumberOfRequiredBuffers());
+			assertEquals("There should be 0 buffer available in local pool", 0, bufferPool.getNumberOfAvailableMemorySegments());
+
+			// Recycle the floating buffer and assign it to the buffer listener
+			floatingBuffer1.recycle();
+
+			// The channel is still waiting for one more floating buffer
+			assertEquals("There should be 12 buffers available in the channel", 12, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 13 buffers required in the channel", 13, inputChannel.getNumberOfRequiredBuffers());
+			assertEquals("There should be 0 buffer available in local pool", 0, bufferPool.getNumberOfAvailableMemorySegments());
+
+			// Recycle one more floating buffer again
+			floatingBuffer2.recycle();
+
+			// The channel already gets all the required buffers
+			assertEquals("There should be 13 buffers available in the channel", 13, inputChannel.getNumberOfAvailableBuffers());
+			assertEquals("There should be 13 buffers required in the channel", 13, inputChannel.getNumberOfRequiredBuffers());
+			assertEquals("There should be 0 buffer available in local pool", 0, bufferPool.getNumberOfAvailableMemorySegments());
+
+			// Decrease the backlog and recycle one exclusive buffer
```

Comment:
please make this two separate tests, i.e. check the invariants after decreasing the backlog size (nothing should change here yet) and once again after recycling an exclusive buffer (possibly integrated with the tests I requested above)